library/core/src/alloc/global.rs

```diff
@@ -53,6 +53,26 @@ use crate::ptr;
 /// * `Layout` queries and calculations in general must be correct. Callers of
 ///   this trait are allowed to rely on the contracts defined on each method,
 ///   and implementors must ensure such contracts remain true.
+///
+/// * You may not rely on allocations actually happening, even if there are explicit
+///   heap allocations in the source. The optimizer may detect allocation/deallocation
+///   pairs that it can instead move to stack allocations/deallocations and thus never
```

Comment:
As I said before, it's not just about alloc/dealloc pairs. Can you word this more generally to talk about "unused allocations" or so? And then note that an allocation can become unused if the compiler can prove that it can use stack memory instead (because it can see all the reads).